### PR TITLE
Caught more mod failure conditions, removed certain assumptions from map creation

### PR DESCRIPTION
**Pull request by yairm210**
My 'example base ruleset', which I used to perform these tests, is now as minimal as I can make it
I plan to make it a template repo so people can start making base rulesets less painfully than has been the case so far

### CODE REVIEW

**Comment by SomeTroglodyte** [comment]
> template repo

_excellent_ idea

> able to mark a difficulty as 'default AI difficulty' somehow

`ModConstants.defaultDifficultyForAI` or `Difficulty.isDefaultForAI` - both would need checker code, first to disallow random text, second to disallow >1 `true`s..

**Comment by yairm210** [comment]
https://github.com/yairm210/example-base-ruleset

This is the boi

File: core/src/com/unciv/logic/civilization/CivilizationInfo.kt [hunk 1]
```diff
@@ -266,7 +266,10 @@ class CivilizationInfo {
     //region pure functions
     fun getDifficulty(): Difficulty {
         if (isPlayerCivilization()) return gameInfo.getDifficulty()
-        return gameInfo.ruleSet.difficulties["Chieftain"]!!
+        // TODO We should be able to mark a difficulty as 'default AI difficulty' somehow
```
Comment (xlenstra) [inline core/src/com/unciv/logic/civilization/CivilizationInfo.kt:269]: ModConstants maybe?

File: core/src/com/unciv/models/ruleset/Ruleset.kt [hunk 2]
```diff
@@ -604,7 +611,7 @@ class Ruleset {
             for (building in era.settlerBuildings)
                 if (building !in buildings)
                     lines += "Nonexistent building $building built by settlers when starting in ${era.name}"
-            if (era.startingMilitaryUnit !in units)
+            if (era.startingMilitaryUnitCount != 0 && era.startingMilitaryUnit !in units)
```
Comment (xlenstra) [inline core/src/com/unciv/models/ruleset/Ruleset.kt:614]: I don't think this catches all uses, as they can also be used in `difficulties.json`

Comment (xlenstra) [inline core/src/com/unciv/models/ruleset/Ruleset.kt:614]: I don't think this catches all uses, as `Difficulties.json` can also use the Era Starting Unit

File: core/src/com/unciv/logic/civilization/CivilizationInfo.kt (hunk 1 above)
Comment (yairm210) [reply core/src/com/unciv/logic/civilization/CivilizationInfo.kt:269]: Perhaps, that's for a different PR though :)

File: core/src/com/unciv/models/ruleset/Ruleset.kt [hunk 3]
```diff
@@ -611,7 +611,7 @@ class Ruleset {
             for (building in era.settlerBuildings)
                 if (building !in buildings)
                     lines += "Nonexistent building $building built by settlers when starting in ${era.name}"
-            if (era.startingMilitaryUnitCount != 0 && era.startingMilitaryUnit !in units)
+            if (era.startingMilitaryUnitCount != 0 && era.startingMilitaryUnit != Constants.eraSpecificUnit && era.startingMilitaryUnit !in units)
```
Comment (xlenstra) [inline core/src/com/unciv/models/ruleset/Ruleset.kt:614]: I think this check is redundant? If we need to add `Constants.eraSpecificUnit`, it is actualy replaced with era.startingMilitaryUnit?

Comment (Hello200000) [reply core/src/com/unciv/models/ruleset/Ruleset.kt:614]: thanks for the warning

Comment (yairm210) [reply core/src/com/unciv/models/ruleset/Ruleset.kt:614]: Then I don't understand your previous comment.

Comment (xlenstra) [reply core/src/com/unciv/models/ruleset/Ruleset.kt:614]: What I meant is that only checking whether the era specific unit exists if that unit is used in the era is incorrect, as it is possible that it is not used in the era's, but is used in by AI's on higher difficulties, meaning that also in those cases the unit should still exist

**Review by xlenstra — APPROVED** [review]

**Pull request merged**